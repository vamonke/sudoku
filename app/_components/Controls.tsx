import { useEffect } from "react";

type Props = {
  restart: () => void;
  newGame: () => void;
  undo: () => void;
  toggleShowConflicts: () => void;
  showConflicts: boolean;
};

export default function Controls(props: Props) {
  const { restart, newGame, undo, toggleShowConflicts, showConflicts } = props;

  useEffect(() => {
    const eventListener = (event: KeyboardEvent) => {
      if (event.key === "r") restart(); // TODO: Add confirmation
      if (event.key === "n") newGame(); // TODO: Add confirmation
      if (event.key === "w") toggleShowConflicts(); // TODO: Add confirmation
      if (event.key === "z") undo();
    };
    window.addEventListener("keyup", eventListener);
    return () => {
      window.removeEventListener("keyup", eventListener);
    };
  }, [newGame, restart, toggleShowConflicts, undo]);

  return (
    <div className="flex flex-row gap-4 mt-2 justify-center items-center">
      <button data-testid="restart" onClick={restart}>
        Restart (R)
      </button>
      <button data-testid="new" onClick={newGame}>
        New Game (N)
      </button>
      <button data-testid="undo" onClick={undo}>
        Undo (Z)
      </button>
      <button data-testid="toggle" onClick={toggleShowConflicts}>
        {showConflicts ? "Hide warnings (W)" : "Show warnings (W)"}
      </button>
    </div>
  );
}
