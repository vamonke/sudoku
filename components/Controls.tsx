import { useEffect } from "react";

type Props = {
  reset: () => void;
  newGame: () => void;
  undo: () => void;
  toggleShowConflict: () => void;
  showConflict: boolean;
};

export default function Controls(props: Props) {
  const { reset, newGame, undo, toggleShowConflict, showConflict } = props;

  useEffect(() => {
    const eventListener = (event: KeyboardEvent) => {
      if (event.key === "r") reset(); // TODO: Add confirmation
      if (event.key === "n") newGame(); // TODO: Add confirmation
      if (event.key === "w") toggleShowConflict();
      if (event.key === "z") undo();
    };
    window.addEventListener("keyup", eventListener);
    return () => {
      window.removeEventListener("keyup", eventListener);
    };
  }, [newGame, reset, toggleShowConflict, undo]);

  return (
    <div className="flex flex-row gap-4 mt-2 justify-center items-center">
      <button data-testid="reset" onClick={reset}>
        Reset (R)
      </button>
      <button data-testid="new" onClick={newGame}>
        New Game (N)
      </button>
      <button data-testid="undo" onClick={undo}>
        Undo (Z)
      </button>
      <button data-testid="toggle" onClick={toggleShowConflict}>
        {showConflict ? "Hide warnings (W)" : "Show warnings (W)"}
      </button>
    </div>
  );
}
