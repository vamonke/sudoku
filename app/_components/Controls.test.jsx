import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import Controls from "./Controls";

describe("Controls", () => {
  it("renders correctly", () => {
    const { container } = render(<Controls />);
    expect(container).toMatchSnapshot();
  });

  it("renders 4 buttons", () => {
    render(<Controls />);
    const buttons = screen.getAllByRole("button");
    expect(buttons).toHaveLength(4);
  });

  it("calls the restart function when the restart button is clicked", () => {
    const mockRestart = jest.fn();
    render(<Controls restart={mockRestart} />);
    const restartButton = screen.getByTestId("restart");
    restartButton.click();
    expect(mockRestart).toHaveBeenCalled();
  });

  it("calls the undo function when the undo button is clicked", () => {
    const mockNewGame = jest.fn();
    render(<Controls undo={mockNewGame} />);
    const undoButton = screen.getByTestId("undo");
    undoButton.click();
    expect(mockNewGame).toHaveBeenCalled();
  });

  it("calls the toggleShowConflict function when the toggle button is clicked", () => {
    const mockUndo = jest.fn();
    render(<Controls toggleShowConflict={mockUndo} />);
    const toggleButton = screen.getByTestId("toggle");
    toggleButton.click();
    expect(mockUndo).toHaveBeenCalled();
  });

  it("calls the newGame function when the new button is clicked", () => {
    const mockNewGame = jest.fn();
    render(<Controls newGame={mockNewGame} />);
    const newButton = screen.getByTestId("new");
    newButton.click();
    expect(mockNewGame).toHaveBeenCalled();
  });

  it("displays the correct text when showConflict is true", () => {
    render(<Controls showConflict={true} />);
    const toggleButton = screen.getByTestId("toggle");
    expect(toggleButton).toHaveTextContent("Hide warnings (W)");
  });

  it("displays the correct text when showConflict is false", () => {
    render(<Controls showConflict={false} />);
    const toggleButton = screen.getByTestId("toggle");
    expect(toggleButton).toHaveTextContent("Show warnings (W)");
  });

  it("should call the restart function when r key is pressed", () => {
    const mockRestart = jest.fn();
    const { container } = render(<Controls restart={mockRestart} />);
    fireEvent.keyUp(container, { key: "r" });
    expect(mockRestart).toHaveBeenCalled();
  });

  it("should call the newGame function when n key is pressed", () => {
    const mockNewGame = jest.fn();
    const { container } = render(<Controls newGame={mockNewGame} />);
    fireEvent.keyUp(container, { key: "n" });
    expect(mockNewGame).toHaveBeenCalled();
  });

  it("should call the undo function when z key is pressed", () => {
    const mockUndo = jest.fn();
    const { container } = render(<Controls undo={mockUndo} />);
    fireEvent.keyUp(container, { key: "z" });
    expect(mockUndo).toHaveBeenCalled();
  });

  it("should call the toggleShowConflict function when w key is pressed", () => {
    const mockToggleShowConflict = jest.fn();
    const { container } = render(
      <Controls toggleShowConflict={mockToggleShowConflict} />
    );
    fireEvent.keyUp(container, { key: "w" });
    expect(mockToggleShowConflict).toHaveBeenCalled();
  });
});
