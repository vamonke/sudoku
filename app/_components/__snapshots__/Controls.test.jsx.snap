// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders correctly 1`] = `
<div>
  <div
    class="flex flex-row gap-4 mt-2 justify-center items-center"
  >
    <button
      data-testid="restart"
    >
      Restart (R)
    </button>
    <button
      data-testid="new"
    >
      New Game (N)
    </button>
    <button
      data-testid="undo"
    >
      Undo (Z)
    </button>
    <button
      data-testid="toggle"
    >
      Show warnings (W)
    </button>
  </div>
</div>
`;
